File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.py
from collections import deque

class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
      # Build graph
      graph = {}
      for i in range(len(edges)):
        a,b = edges[i]
        edge_prob = succProb[i]
        if a not in graph:
          graph[a] = []
        if b not in graph:
          graph[b] = []
        graph[a].append([b,edge_prob])
        graph[b].append([a,edge_prob])
      
      
      prob_arr = [0]*n
      prob_arr[start] = 1
      queue = collections.deque([[start, 1]])
      
      while(len(queue) > 0): 
        node,curr_prob = queue.popleft()
        if node in graph:
          for neighbor,neigh_prob in graph[node]:
            if curr_prob*neigh_prob > prob_arr[neighbor]:
              prob_arr[neighbor] = curr_prob*neigh_prob
              queue.append([neighbor,prob_arr[neighbor]])
      
      return prob_arr[end]

File: 1514-path-with-maximum-probability/1514-path-with-maximum-probability.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number[]} succProb
 * @param {number} start
 * @param {number} end
 * @return {number}
 */
var maxProbability = function(n, edges, succProb, start, end) {
  // Undirected, weighted graph
  const graph = buildGraph(edges, succProb);

  // probability array for each node
  const prob = Array(n).fill(0);
  const queue = [[start, 1]];
  prob[start] = 1;

  while (queue.length > 0) {
    let [node, currProb] = queue.shift();
    if (graph[node]) {
      for (let [neighbor, nextProb] of graph[node]) {
        if (currProb * nextProb > prob[neighbor]) {
          prob[neighbor] = currProb * nextProb;
          queue.push([neighbor, prob[neighbor]])
        }
      }
    }
  }
  
  return prob[end];
};




const buildGraph = (edges, succProb) => {
  const graph = {};
  for (let i = 0; i < edges.length; i++) {
    const [a,b] = edges[i];
    const prob = succProb[i];
    if (!graph[a]) graph[a] = [];
    if (!graph[b]) graph[b] = [];
    graph[a].push([b, prob]);
    graph[b].push([a, prob]);
  }
  return graph;
}